### PR TITLE
CATTY-449 New Formula Editor screen

### PR DESCRIPTION
**Pull request by neel-makhecha**
Implemented the new FormulaEditorSectionViewController to conveniently select functions with subsections, and select, add or delete variables and lists.
Implemented new FormulaEditorSubsection protocol and updated the FormulaEditorSection enum to support subsections.
Updated all the functions, operators, etc. to categorise in new sections and subsections.

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catty/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [ ] Include the name of the Jira ticket in the PR’s title
- [ ] Verify that the Jira ticket is in the status *Ready for Development*
- [ ] Include a summary of the changes plus the relevant context
- [ ] Choose the proper base branch (*develop*)
- [ ] Confirm that the changes follow the project’s coding guidelines
- [ ] Verify that the changes generate no compiler or linter warnings
- [ ] Perform a self-review of the changes
- [ ] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [ ] Confirm that new and existing unit tests pass locally
- [ ] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [ ] Stick to the project’s git workflow (rebase and squash your commits)
- [ ] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *#catty* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewControllerDataExtension.swift [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+            self.spriteObject?.userData.add(userList)
+        }
+
+        self.spriteObject?.scene.project?.saveToDisk(withNotification: false)
+        self.reloadData()
+    }
+
+    func isVariableUsed(_ userVariable: UserVariable) -> Bool {
+
+        guard let project = self.spriteObject?.scene.project else {
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewControllerDataExtension.swift:146]: Could you please come up with a unit test which is verifying the behavior of `isVariableUsed` and `isListUsed`? Just make sure that this behavior is not tested using a UI test. Once we have a regression test, I am sure we can do some refactoring here.

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+            if let groupedItems = dict[subsection] {
+                items.append(contentsOf: groupedItems)
+                sizes.append(groupedItems.count)
+            }
+        }
+
+        return sizes
+    }
+
+    private func initDataItems() {
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift:321]: There still seems to be a lot of code which is solely for the data section. So I was wondering: would it be more readable if we create a new VC class for this section (could be derived from the FESectionVC). Maybe this could also be an idea for all other sections (e.g., in order to get rid of the RawValue enum workaround). What do you think?

File: src/Catty/FormulaEditor/ViewController/FormulaEditorViewController.m [hunk 3]
```diff
@@ -272,13 +254,19 @@ - (void)viewDidLoad
     
     panRecognizer.delegate = self;
     
-    [self.variablePicker addGestureRecognizer:tapToSelect];
-    [self.variablePicker addGestureRecognizer:panRecognizer];
+    [self setupNavigationBar];
+    
+    self.formulaEditorSectionViewController = [[FormulaEditorSectionViewController alloc] init];
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorViewController.m:259]: Is the `FormulaEditorSectionViewController` also working without the `FormulaManager` and `SpriteObject`? If not, please add them directly add the initializer and make them non-optional, so that every instance must have those necessary properties assigned.

File: src/Catty/FormulaEditor/ViewController/FormulaEditorViewController.m [hunk 4]
```diff
@@ -569,7 +542,7 @@ - (void)showFormulaEditorTextView
 {
     CGFloat marginTop = self.brickCell.frame.origin.y + self.brickCell.frame.size.height;
     
-    self.formulaEditorTextView = [[FormulaEditorTextView alloc] initWithFrame: CGRectMake(0, marginTop, self.view.frame.size.width - 2, 120) AndFormulaEditorViewController:self];
+    self.formulaEditorTextView = [[FormulaEditorTextView alloc] initWithFrame: CGRectMake(0, marginTop, self.view.frame.size.width, 120) AndFormulaEditorViewController:self];
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorViewController.m:545]: Please make the text view a little bit larger again, i.e., put back the `-2`.

File: src/CattyUITests/ScriptCollectionView/ListTests.swift [hunk 5]
```diff
@@ -71,56 +71,6 @@ class ListTests: XCTestCase {
     }
 
     func testCreateAndSelectList() {
-        let listName = "TestList"
```
Comment (m-herold) [inline src/CattyUITests/ScriptCollectionView/ListTests.swift:74]: Is there a special reason why you removed those existing tests? Are they not used anymore?

File: src/CattyUITests/ScriptCollectionView/VariableTests.swift [hunk 6]
```diff
@@ -72,59 +72,15 @@ class VariableTests: XCTestCase {
     }
 
     func testCreateAndSelectVariable() {
-        let variableName = "testVariable"
```
Comment (m-herold) [inline src/CattyUITests/ScriptCollectionView/VariableTests.swift:75]: Same question here: why are those tests not relevant anymore?

**Review by m-herold — CHANGES_REQUESTED** [review]

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+        let tableViewLeadingConstraint = NSLayoutConstraint(item: self.tableView, attribute: .leading, relatedBy: .equal, toItem: self.view, attribute: .leading, multiplier: 1, constant: 0)
+        let tableViewTrailingConstraint = NSLayoutConstraint(item: self.tableView, attribute: .trailing, relatedBy: .equal, toItem: self.view, attribute: .trailing, multiplier: 1, constant: 0)
+        self.tableView.translatesAutoresizingMaskIntoConstraints = false
+        self.tableView.dataSource = self
+        self.tableView.delegate = self
+        self.view.addSubview(self.tableView)
+        self.view.addConstraints([tableViewTopConstraint, tableViewBottomConstraint, tableViewLeadingConstraint, tableViewTrailingConstraint])
+
+        self.placeHolderLabel = UILabel(frame: CGRect(width: self.view.frame.width, height: 45))
+        self.placeHolderLabel.text = kUIFEAddVariablesAndLists
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift:80]: Can you please make sure that this text is also fully visible on smaller devices (e.g., iPhone SE models). Like on the Object-screen it should be then displayed in multiple lines (in the case the screen is too small):

Current situation:
Data

Desired situation:
Objects

**Review by m-herold — CHANGES_REQUESTED** [review]

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewControllerDataExtension.swift (hunk 1 above)
Comment (neel-makhecha) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewControllerDataExtension.swift:146]: We already have `FormulaEditorSectionViewControllerTests` where `isVariableUsed` and `isListUsed` is tested. Is that what you are asking or am I misunderstanding?

File: src/CattyUITests/ScriptCollectionView/ListTests.swift (hunk 5 above)
Comment (neel-makhecha) [reply src/CattyUITests/ScriptCollectionView/ListTests.swift:74]: Yes, they are not used anymore. Because we don't have a "choose" button anymore.

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift (hunk 2 above)
Comment (neel-makhecha) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift:321]: Thanks! This sounds awesome! 
PS: I also did something like this for the keyboard buttons that you'll see in CATTY-517!

Comment (neel-makhecha) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift:321]: However, I think this should only be done for the data section, as behaviour of all other sections is identical. But in that case we will still need enums. What's your suggestion on that?

Comment (m-herold) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift:321]: Perfect, the data section is also the most important one for me. The others do differ in the initialization which is also not negligible and would be enough for me to create a new class, but that must not be. Please only go for the data section at the moment. Thanks!

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewControllerDataExtension.swift (hunk 1 above)
Comment (m-herold) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewControllerDataExtension.swift:146]: Perfect! So let's start refactoring: I would suggest to get rid of a lot of duplicated code by introducing the following two methods:

`- private func getAllBricks(for object: SpriteObject) -> [Brick]`

This method could be used for checking variables/lists with both global and local scope.

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewControllerDataExtension.swift [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+            fatalError("project of the spriteObject is nil")
+        }
+
+        if project.userData.contains(userVariable) {
+
+            if let objects = self.spriteObject?.scene.objects() {
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewControllerDataExtension.swift:152]: Please use `project.allObjects()` instead to get all objects. That's necessary because later we could have several scenes.

**Review by m-herold — CHANGES_REQUESTED** [review]

**Comment by neel-makhecha** [comment]
Hi Michael, I have pushed a new commit to this PR with all the changes you suggested above. Please review it.

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift (hunk 2 above)
Comment (neel-makhecha) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift]: I agree, separate classes for each section would be a better approach than enums. Please review the new `FormulaEditorDataSectionViewController`, if it looks good, I shall implement similar classes for all other sections too.

File: src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift [hunk 9]
```diff
@@ -87,27 +54,6 @@ extension FormulaEditorViewController {
...
-            button.topAnchor.constraint(equalTo: previousButton?.bottomAnchor ?? scrollView.topAnchor, constant: 0).isActive = true
-        } else if let topAnchorView = previousButton {
-            button.leftAnchor.constraint(equalTo: topAnchorView.rightAnchor, constant: 0).isActive = true
-            button.topAnchor.constraint(equalTo: topAnchorView.topAnchor, constant: 0).isActive = true
-        }
-
-        return button
-    }
-
     private func resizeSection(scrollView: UIScrollView, for buttons: [UIButton], with buttonHeight: CGFloat, fullButtonWidth: Bool) {
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift:57]: Do we still need this function? I think it is unused by now.

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController/FormulaEditorDataSectionViewController.swift [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+@objc class FormulaEditorDataSectionViewController: FormulaEditorSectionViewController {
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController/FormulaEditorDataSectionViewController.swift:23]: Thanks for introducing this class. This makes the original `FormulaEditorSectionViewController` much more compact. Suggestion: now the data extension for the base class is not really useful anymore (`FormulaEditorSectionViewControllerDataExtension`). How about moving it into that view controller directly?

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController/FormulaEditorDataSectionViewController.swift [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+    private var addButton = UIBarButtonItem()
+    private var placeHolderLabel = UILabel()
+
+    var variableSourceProject = [UserVariable]()
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController/FormulaEditorDataSectionViewController.swift:28]: Could you please check if those variables really need to be public?

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController/FormulaEditorSectionViewControllerDataExtension.swift [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+extension FormulaEditorDataSectionViewController {
+
+    func askProjectOrObject(isList: Bool) {
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController/FormulaEditorSectionViewControllerDataExtension.swift:25]: Please see my comments above for this extension.

File: src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift [hunk 13]
```diff
@@ -22,49 +22,16 @@
...
-    }
-
-    private func initWithItems(formulaEditorItems: [FormulaEditorItem], scrollView: UIScrollView, buttonHeight: CGFloat, fullButtonWidth: Bool) -> [UIButton] {
-        var button: UIButton?
-        var buttons = [UIButton]()
-
-        for item in formulaEditorItems {
-            button = buttonForScrollView(item: item, scrollView: scrollView, previousButton: button, buttonHeight: buttonHeight, fullButtonWidth: fullButtonWidth)
-            buttons.append(button!)
+    func formulaEditorItemSelected(item: FormulaEditorItem) {
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift:25]: How about integrating these decisions into `FormulaEditorSectionViewController` directly? Now that the view controller is very small and since that is still part of the section, we could integrate it there directly. The public interface to the `FormulaEditor` would then be `handleInput` which would be very readable and clean. And very good to be tested using unit test and test doubles (please add a unit test for that).

**Review by m-herold — CHANGES_REQUESTED** [review]

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift (hunk 2 above)
Comment (m-herold) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift]: Please see my comments directly at the files. Both approaches are fine for me -- if you keep the enums, could you please get rid of the `.none` enum?

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController/FormulaEditorSectionViewController.swift [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+    case functions
+    case object
+    case logic
+    case sensors
+}
+
+@objc class FormulaEditorSectionViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
+
+    @objc var formulaEditorSectionType: FormulaEditorSectionType = .none
+    @objc var formulaManager: FormulaManager
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController/FormulaEditorSectionViewController.swift:34]: I think those three variables (`formulaManager`, `spriteObject` and `formulaEditorVC`) are not supposed to change, right? If so, could you please change to `let`?

**Review by m-herold — CHANGES_REQUESTED** [review]

File: src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift (hunk 13 above)
Comment (neel-makhecha) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift]: Hi Michael, I have pushed a new commit to this PR with all the changes you suggested, including this one. However, will you please give some suggestions on how exactly should I write unit tests for this? Also, whether I have understood and implemented this properly.

Comment (m-herold) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift]: I have added an example test case for `FormulaEditorSectionViewControllerTests`. If you agree with this test, could you please add it for the other sections as well?

File: src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift [hunk 15]
```diff
@@ -22,51 +22,6 @@
 
 extension FormulaEditorViewController {
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift:23]: Since this extension is not directly related to sections, my suggestion would be to rename it to `FormulaEditorViewControllerInputExtension` or similar. What do you think?

**Review by m-herold — CHANGES_REQUESTED** [review]

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController/FormulaEditorSectionViewController.swift [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+        } else if let function = item.function {
+            self.formulaEditorVC.handleInput(for: function)
+        } else if let op = item.op {
+            self.formulaEditorVC.handleInput(for: op)
+        }
+
+    }
+
+    func presentUnexpectedErrorAlert() {
+        let alert = UIAlertController(title: "Some unexpected error occured", message: nil, preferredStyle: .alert)
```
Comment (m-herold) [inline src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController/FormulaEditorSectionViewController.swift:218]: Please replace hardcoded translation with `kLocalizedUnexpectedErrorTitle` and `kLocalizedUnexpectedErrorMessage`.

**Review by m-herold — CHANGES_REQUESTED** [review]

File: src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift (hunk 15 above)
Comment (neel-makhecha) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift:23]: Yes, I have renamed that file to `FormulaEditorViewControllerInputExtension`.

File: src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift (hunk 13 above)
Comment (neel-makhecha) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorViewController+Section.swift]: Thank you very much! I have pushed a new commit with tests and multiple other changes. Please review it.

File: src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift (hunk 2 above)
Comment (neel-makhecha) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift]: I had kept `.none` case inside this enum because it was serving for the "gap" by the inheriting classes (such as, FormulaEditorDataSectionViewController). Additionally, I can't have that property as an optional, because optional enums are not represented in Objective-C. So I have added back the `.data` case again, and updated the initialisers. However, I am not sure whether this is the best solution. What do you think?

Comment (m-herold) [reply src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewController.swift]: Thanks for the explanation! I agree that it is not perfect, but also that it is not easily solvable in a better way with the current architecture. For me this approach is fine.

File: src/CattyTests/ViewController/FormulaEditorSectionViewControllerTests.swift [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+
+                assertThatInputIsValid(for: internFormula, having: expectedItems)
+                itemsInTableView += 1
+            }
+        }
+
+        XCTAssertEqual(expectedItems.count, itemsInTableView)
+    }
+
+    func testInitAndSelectLogicSection() {
```
Comment (m-herold) [inline src/CattyTests/ViewController/FormulaEditorSectionViewControllerTests.swift:69]: Thanks! Do make the tests complete, please add one UI test each for selecting a logic item, a function, an object property and a sensor. You can leave out the data section, since this is already tested thoroughly.

**Review by m-herold — CHANGES_REQUESTED** [review]

File: src/CattyTests/ViewController/FormulaEditorSectionViewControllerTests.swift (hunk 17 above)
Comment (neel-makhecha) [reply src/CattyTests/ViewController/FormulaEditorSectionViewControllerTests.swift:69]: Hi Michael, I have pushed a new commit with these UI tests. Please review it.

**Review by m-herold — APPROVED** [review]
Great work, thanks!
